Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -1492,101 +1493,202 @@ google_firestore_v1_MapValue Serializer::EncodeMapValue(
+      return "limbo-document";
+    default:
+      HARD_FAIL("Unrecognized query purpose: %s", purpose);
+  }
 }
 
 std::unique_ptr<WatchChange> Serializer::DecodeWatchChange(
     nanopb::Reader* reader,
     const google_firestore_v1_ListenResponse& watch_change) const {
-  // TODO(varconst): implement.
-  // - (std::unique_ptr<WatchChange>)decodedWatchChange:
-  //     (GCFSListenResponse *)watchChange;
-  return {};
+  switch (watch_change.which_response_type) {
+    case google_firestore_v1_ListenResponse_target_change_tag:
+      return DecodeTargetChange(reader, watch_change.target_change);
+
+    case google_firestore_v1_ListenResponse_document_change_tag:
+      return DecodeDocumentChange(reader, watch_change.document_change);
+
+    case google_firestore_v1_ListenResponse_document_delete_tag:
+      return DecodeDocumentDelete(reader, watch_change.document_delete);
+
+    case google_firestore_v1_ListenResponse_document_remove_tag:
+      return DecodeDocumentRemove(reader, watch_change.document_remove);
+
+    case google_firestore_v1_ListenResponse_filter_tag:
+      return DecodeExistenceFilterWatchChange(reader, watch_change.filter);
+
+    default:
+      HARD_FAIL("Unknown WatchChange.changeType %s",
+                watch_change.which_response_type);
+  }
 }
 
 SnapshotVersion Serializer::DecodeVersion(
     nanopb::Reader* reader,
     const google_firestore_v1_ListenResponse& listen_response) const {
-  // TODO(varconst): implement.
-  // - (SnapshotVersion)versionFromListenResponse:
-  //     (GCFSListenResponse *)watchChange;
-  return {};
+  // We have only reached a consistent snapshot for the entire stream if there
+  // is a read_time set and it applies to all targets (i.e. the list of targets
+  // is empty). The backend is guaranteed to send such responses.
+  if (listen_response.which_response_type !=
+      google_firestore_v1_ListenResponse_target_change_tag) {
+    return SnapshotVersion::None();
+  }
+  if (listen_response.target_change.target_ids_count != 0) {
+    return SnapshotVersion::None();
+  }
+  return DecodeSnapshotVersion(reader, listen_response.target_change.read_time);
+}
+
+std::vector<TargetId> Serializer::DecodeTargetIdArray(nanopb::Reader* reader,
+                                                      int32_t* array,
+                                                      pb_size_t size) const {
+  std::vector<TargetId> target_ids;
+
+  for (pb_size_t i = 0; i < size; i++) {
+    target_ids.push_back(array[i]);
+  }
+  return target_ids;
 }
 
 std::unique_ptr<WatchChange> Serializer::DecodeTargetChange(
     nanopb::Reader* reader,
     const google_firestore_v1_TargetChange& change) const {
-  // TODO(varconst): implement.
-  // - (std::unique_ptr<WatchChange>)decodedTargetChangeFromWatchChange:
-  //     (GCFSTargetChange *)change;
-  return {};
+  WatchTargetChangeState state =
+      DecodeTargetChangeState(reader, change.target_change_type);
+  auto target_ids =
```

Comment:
Nit: in C++, if you use the `vector` constructor that takes two iterators, then this helper function becomes pretty unnecessary:
```cpp
std::vector<TargetId> target_ids(change.target_ids, change.target_ids + change.target_ids_count);
```